kernel/arch/dreamcast/kernel/init.c

```diff
@@ -231,10 +231,15 @@ void  __weak arch_auto_shutdown(void) {
     if (!KOS_PLATFORM_IS_NAOMI)
         KOS_INIT_FLAG_CALL(net_shutdown);
 
-    irq_disable();
     snd_shutdown();
-    timer_shutdown();
     hardware_shutdown();
+    /* XXX: We should investigate shrinking this irq_disabled
+       time. Until then, all these shut downs happen with
+       irqs disabled which prevents things like safely joining
+       threads or sending cleanup commands to hardware.
+    */
+    irq_disable();
+    timer_shutdown();
```

Comment:
As in the note, shrinking this irq_disabled to ensure our shutdowns are not limited by that is valuable in itself. Here though, we could not properly shut down threads for the maple callbacks using `thd_join` with irqs disabled. Doing so would just send the callback threads into a genwait_wait that it would never return from (as there was no more context switching). `timer_shutdown` had to be moved for the same reason, as turning off the timers would also prevent context switching and cause a `genwait_wait` to be an infinite loop.

None of the other shutdowns in `hardware_shutdown` should be impacted, as most do little or nothing. Also `net_shutdown` was already happening prior to the irq disable in order to be able to clean up the net thread.